Fixing issue with multiple collect call on Magento\GiftCardAccount\Model\Total\Quote\Giftcardaccount which removes giftcard from an order (DOG-715)
This process is described here https://github.com/magento/magento2/issues/30029
Logic of collet method of Giftcardaccount causes that on each second call giftcard is removed form the order.

diff --git a/Model/Total/Quote/Giftcardaccount.php b/Model/Total/Quote/Giftcardaccount.php
index f9d5fac..8355081 100644
--- a/Model/Total/Quote/Giftcardaccount.php
+++ b/Model/Total/Quote/Giftcardaccount.php
@@ -31,6 +31,8 @@ class Giftcardaccount extends \Magento\Quote\Model\Quote\Address\Total\AbstractT
      */
     protected $priceCurrency;

+    private $isCollected = false;
+
     /**
      * @param \Magento\GiftCardAccount\Helper\Data $giftCardAccountData
      * @param \Magento\GiftCardAccount\Model\GiftcardaccountFactory $giftCAFactory
@@ -62,6 +64,10 @@ class Giftcardaccount extends \Magento\Quote\Model\Quote\Address\Total\AbstractT
         \Magento\Quote\Api\Data\ShippingAssignmentInterface $shippingAssignment,
         \Magento\Quote\Model\Quote\Address\Total $total
     ) {
+        if($this->isCollected) {
+            return $this;
+        }
+
         $baseAmountLeft = $quote->getBaseGiftCardsAmount() - $quote->getBaseGiftCardsAmountUsed();
         $amountLeft = $quote->getGiftCardsAmount() - $quote->getGiftCardsAmountUsed();

@@ -147,6 +153,8 @@ class Giftcardaccount extends \Magento\Quote\Model\Quote\Address\Total\AbstractT
         $total->setBaseGiftCardsAmount($baseUsed);
         $total->setGiftCardsAmount($used);

+        $this->isCollected = true;
+
         return $this;
     }

